File: backend/apps/tickets/validators/file_validator.py
"""
Chain of Responsibility node — validates uploaded file attachments.

Responsibility (SRP): enforce only file size and MIME type rules.
Depends on: BaseValidator (core/base/base_validator.py).
Pattern: Chain of Responsibility node.
SOLID: SRP · OCP · LSP
"""

from __future__ import annotations

from core.base.base_validator import BaseValidator, ValidationResult


class FileValidator(BaseValidator):
    """Validates file size (≤5 MB) and MIME type for ticket attachments."""

    MAX_SIZE_BYTES: int = 5_242_880  # 5 MB

    ALLOWED_MIME_TYPES: frozenset[str] = frozenset({
        "image/jpeg",
        "image/png",
        "image/gif",
        "application/pdf",
        "application/msword",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "text/plain",
    })

    def validate(self, data: dict) -> ValidationResult:
        adjuntos = data.get("adjuntos", [])
        if not adjuntos:
            return ValidationResult(is_valid=True)

        for file in adjuntos:
            if file.size > self.MAX_SIZE_BYTES:
                return ValidationResult(
                    is_valid=False,
                    errors=[
                        f"'{file.name}' excede el límite de 5 MB "
                        f"({file.size / 1_048_576:.1f} MB)."
                    ],
                    field_name="adjuntos",
                )
            if file.content_type not in self.ALLOWED_MIME_TYPES:
                return ValidationResult(
                    is_valid=False,
                    errors=[
                        f"Tipo de archivo '{file.content_type}' no está permitido. "
                        f"Tipos aceptados: PDF, Word, imágenes y texto plano."
                    ],
                    field_name="adjuntos",
                )

        return ValidationResult(is_valid=True)

File: frontend/src/modules/tickets/validators/FileValidator.ts
/**
 * Chain of Responsibility node — validates file attachments on the ticket form.
 *
 * Responsibility (SRP): enforce only file size and MIME type rules.
 *     No text field checks, no API calls — just file constraints.
 * Depends on: BaseValidator (src/core/base/BaseValidator.ts).
 * Pattern: Chain of Responsibility node.
 * SOLID: SRP · OCP · LSP
 *
 * Rules enforced:
 *   - Each file: size ≤ 5 MB (5_242_880 bytes)
 *   - Each file: MIME type in ALLOWED_MIME_TYPES
 *
 * OCP: new MIME type allowed = add to ALLOWED_MIME_TYPES set; no logic change.
 *      New size policy = new node; this class unchanged.
 */

import { BaseValidator, type ValidationResult } from '../../../core/base/BaseValidator'

export class FileValidator extends BaseValidator {
  private static readonly MAX_SIZE_BYTES = 5_242_880 // 5 MB

  private static readonly ALLOWED_MIME_TYPES = new Set([
    'image/jpeg',
    'image/png',
    'image/gif',
    'application/pdf',
    'application/msword',
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'text/plain',
  ])

  validate(data: unknown): ValidationResult {
    const { adjuntos = [] } = data as { adjuntos?: File[] }
    for (const file of adjuntos) {
      if (file.size > FileValidator.MAX_SIZE_BYTES)
        return { isValid: false, field: 'adjuntos', errors: [`"${file.name}" supera el tamaño máximo de 5 MB.`] }
      if (!FileValidator.ALLOWED_MIME_TYPES.has(file.type))
        return { isValid: false, field: 'adjuntos', errors: [`"${file.name}" tiene un formato no permitido.`] }
    }
    return { isValid: true, field: '', errors: [] }
  }
}
